### PR TITLE
Implementation of TIP parser

### PR DESCRIPTION
**Pull request by delanoerhardt**
Initial implementation, still needs tests and some fixes. I'm having some difficulty with:
- [X] parsing of indirect function calls. I made it so that all function calls use variable expressions, due to stack overflows when using a naive dependency on expressions as a whole;
  - Direct function calls only use IDs, indirect use all expressions, but VariableExp due to it being impossible to distinguish from plain IDs; 
- [X] some details of TIP that seem somewhat ill defined; some grammar rules (ie. store operation) seem weird
  - I was overcomplicating things
- [x] unsure about rules of precedence on operators, but I assume this isn't much of a problem.
- [x] Implement tests for:
  - [X] Expressions;
  - [X] Statments;
  - [x] Functions;
  - [x] Programs.

### CODE REVIEW

**Comment by delanoerhardt** [comment]
The problem with direct and indirect function calls is that the definitions eclipse one another. Indirect function calls allow an expression to be used as an identifier for the function, while direct function calls only allow ids to be used as an identifier.

The problem is that the parsing rules for variable expressions and id is the same by definition. If we activate the directFunctionCallParser every "name(...args)" will be parsed as direct function calls, even when name should point to a variable.

It seems that without two types of conflicting function call expressions this would be solved. The function would always be identified by an expression, and plain ids would be treated as VariableExp which is likely the intention anyway.

What do you think @rbonifacio @Jclavo?

File: src/test/scala/br/unb/cic/tip/parser/StatementParserTest.scala [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import br.unb.cic.tip.*
+
+import br.unb.cic.tip.Stmt.*
+import br.unb.cic.tip.Expression.*
+
+class StatementParserTest extends AnyFunSuite {
+  ignore("should parse empty strings") {
+    // TODO: this is required by the grammar rules.
+    // Is it implementable  with the current AST?
+    val result = StatementParser.parse("")
```
Comment (rbonifacio) [inline src/test/scala/br/unb/cic/tip/parser/StatementParserTest.scala:11]: Good point. I think the AST does not support the empty statement.

**Review by rbonifacio — COMMENTED** [review]
@delanoerhardt . Excellent contribution! Did you find a solution for the two types of "calls"? I think that we could implement only the most general one.

File: src/test/scala/br/unb/cic/tip/parser/StatementParserTest.scala (hunk 1 above)
Comment (delanoerhardt) [reply src/test/scala/br/unb/cic/tip/parser/StatementParserTest.scala:11]: Done! Added a NopStmt to Syntax

**Comment by delanoerhardt** [comment]
@rbonifacio I merged the two function call expressions into one and finally got around to adding the needed tests.

I think this is ready to merge now. Anything that needs to be changed?

**Review by Jclavo — APPROVED** [review]

**Comment by Jclavo** [comment]
@delanoerhardt can you solve the conflicts, pls?

**Pull request merged**